### ISSUE TITLE
Double imports in node makes the lib buggy

### ISSUE THREAD

**Issue by thomas-hilaire** [kewisch/ical.js#279]
Hello,

Having two node modules linked with "npm link" breaks the library when both "require" it.
Here are reproduction steps that prove it (or tell me if I do things in a wrong way):

**Firstly, check that the library works as expected when imported only once:**
```
$ mkdir first-module-with-ical
$ cd first-module-with-ical
$ npm init
$ npm i --save ical.js
```
Then make a "index.js" file that read an ICS then display its startDate:

```
'use strict';

var ICAL = require("ical.js");

var icalendar = `BEGIN:VCALENDAR
VERSION:2.0
PRODID:-//Sabre//Sabre VObject 4.1.2//EN
BEGIN:VTIMEZONE
TZID:Europe/Berlin
BEGIN:DAYLIGHT
TZOFFSETFROM:+0100
TZOFFSETTO:+0200
TZNAME:CEST
DTSTART:19700329T020000
RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=-1SU
END:DAYLIGHT
BEGIN:STANDARD
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
TZNAME:CET
DTSTART:19701025T030000
RRULE:FREQ=YEARLY;BYMONTH=10;BYDAY=-1SU
END:STANDARD
END:VTIMEZONE
BEGIN:VEVENT
UID:514ffe29-f677-4bb9-bf2f-25b038b0c5ba
TRANSP:OPAQUE
DTSTART;TZID=Europe/Berlin:20161220T203000
DTEND;TZID=Europe/Berlin:20161220T213000
ORGANIZER;CN=admin admin:mailto:admin@domain.org
SUMMARY:vvvvvvvvvv
ATTENDEE;PARTSTAT=ACCEPTED;RSVP=FALSE;ROLE=CHAIR:mailto:admin@domain.org
DTSTAMP:20161220T152711Z
END:VEVENT
END:VCALENDAR`;

var vcalendar = ICAL.Component.fromString(icalendar);
var event = vcalendar.getFirstSubcomponent('vevent');
var icalEvent = new ICAL.Event(event);

console.log(icalEvent.startDate);
```

Run it and see the result:
```
$ node index.js 
	{ wrappedJSObject: [Circular],
	  _time: 
	   { year: 2016,
	     month: 12,
	     day: 20,
	     hour: 20,
	     minute: 30,
	     second: 0,
	     isDate: false },
	  _cachedUnixTime: null,
	  _pendingNormalization: true,
	  timezone: 'Europe/Berlin',
	  zone: 
	   { wrappedJSObject: [Circular],
	     expandedUntilYear: 0,
	     changes: [],
	     tzid: 'floating' } }
```

**In a second time, check that the library does not work as expected when imported twice:**

Following the previous steps, make the code depending on another module that is importing ical.js too:
```
$ mkdir ../other-module-with-ical.js
$ cd ../other-module-with-ical.js
$ npm init
$ npm i --save ical.js
$ echo "var ICAL = require('ical.js');" > index.js
$ cd -
$ npm link ../other-module-with-ical.js
$ sed -i 's/require("ical.js");/require("ical.js");require("other-module-with-ical.js");\n/' index.js
```

Run it and see the result, we now have a null "startDate" just because we required twice ical.js :
```
$ node index.js
	null
```

What do you think about this behavior ? Thanks for your work ;)

I'm using node v6.5.0 and ical.js 1.2.2.

**Comment by kewisch** [comment]
Ah yes, this has bugged me often enough when working on performance tests. The way the files are loaded is a bit strange.

There are a few ways to go about this, but all of them are not trivial. To do it right I think it would be required to go through all the files to make sure there are no top-level circular dependencies and change each file to export only the object it describes. Then add a main file that exports the `ICAL` object.

Going even further, the code could be rewritten to use ES6 classes and then use babel/webpack to generate the build files. Also not trivial, but if changing the module loader anyway we might as well change to ES6.

Did you find a workaround by now?

**Comment by thomas-hilaire** [comment]
Hello Philipp!

I only did a really quick workaround to not waste my time before I get your mind of this topic. Something like the following snippet at the beginning of `build/ical.js` :
```
if (typeof ICAL !== 'undefined') {
  module.exports = ICAL;
  return;
}
```

The switch on ES6 classes seems the best choice, and yes that's not a trivial work I guess ;)
But it could be a nice way for us to contribute to your project!

Thanks for your reply

**Comment by kewisch** [comment]
Yeah, I had a similar hack in there for a moment, but this of course fails during performance tests where there is an expectation of having two versions with different code. I certainly don't want to burden you with such a large change (ES6), but of course I welcome the contribution. When you have cycles for this lets open a new issue and discuss the packaging. I'd love to hear if your practices regarding the use of babel have evolved and how you would solve it.

**Comment by kewisch** [comment]
Closing in favor of #470 and an upcoming es6 rewrite
